extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLSubscriptionHandler.java

```diff
@@ -0,0 +1,180 @@
+import graphql.GraphqlErrorBuilder;
+import io.quarkus.security.identity.CurrentIdentityAssociation;
+import io.quarkus.vertx.http.runtime.CurrentVertxRequest;
+import io.smallrye.graphql.bootstrap.Config;
+import io.smallrye.graphql.execution.ExecutionResponse;
+import io.smallrye.graphql.execution.error.ExecutionErrorsService;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.helpers.Subscriptions;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.ServerWebSocket;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Handler that does the execution of GraphQL Requests
+ */
+public class SmallRyeGraphQLSubscriptionHandler extends SmallRyeGraphQLAbstractHandler {
+    private static final Logger log = Logger.getLogger(SmallRyeGraphQLSubscriptionHandler.class);
+    private final ExecutionErrorsService executionErrorsService;
+    private final Config config;
+    private final AtomicReference<Subscription> subscriptionRef = new AtomicReference<>();
+
+    public SmallRyeGraphQLSubscriptionHandler(Config config, CurrentIdentityAssociation currentIdentityAssociation,
+            CurrentVertxRequest currentVertxRequest) {
+        super(currentIdentityAssociation, currentVertxRequest);
+        this.config = config;
+        this.executionErrorsService = new ExecutionErrorsService(config);
+    }
+
+    @Override
+    protected void doHandle(final RoutingContext ctx) {
+        if (ctx.request().headers().contains(HttpHeaders.UPGRADE, HttpHeaders.WEBSOCKET, true) && !ctx.request().isEnded()) {
+            ctx.request().toWebSocket(new SmallRyeWebSocketHandler());
+        } else {
+            ctx.next();
+        }
+    }
+
+    private class SmallRyeWebSocketHandler implements Handler<AsyncResult<ServerWebSocket>> {
+
+        @Override
+        public void handle(AsyncResult<ServerWebSocket> event) {
+            if (event.succeeded()) {
+                ServerWebSocket serverWebSocket = event.result();
+                serverWebSocket.closeHandler(new CloseHandler());
+                serverWebSocket.endHandler(new EndHandler());
+                serverWebSocket.exceptionHandler(new ExceptionHandler());
+                serverWebSocket.textMessageHandler(new TextMessageHandler(serverWebSocket));
+            }
+        }
+    }
+
+    private class CloseHandler implements Handler<Void> {
+        @Override
+        public void handle(Void e) {
+            unsubscribe();
+        }
+    }
+
+    private class EndHandler implements Handler<Void> {
+        @Override
+        public void handle(Void e) {
+            unsubscribe();
+        }
+    }
+
+    private class ExceptionHandler implements Handler<Throwable> {
+        @Override
+        public void handle(Throwable e) {
+            log.error(e.getMessage());
+            unsubscribe();
+        }
+    }
+
+    public void unsubscribe() {
+        Subscriptions.cancel(subscriptionRef);
```

Comment:
Make sure this method checks for null subscription.

Reply:
Done